Ingo 260123 (#52)
IngJohEricsson/draft-johansson-ccwg-scream-bis: draft-johansson-ccwg-rfc8298bis-screamv2.md

sudden delay spikes due to e.g. handover or link layer
retransmissions.

qdelay_avg is updated with a slow attack, fast decay EWMA filter as described below.

~~~
if (now - last_update_qdelay_avg_time >= min(virtual_rtt,s_rtt)
  # Calculate qdelay_avg
  if (qdelay < qdelay_avg)
    qdelay_avg = qdelay
  else
    qdelay_avg = QDELAY_AVG_G*qdelay + (1.0-QDELAY_AVG_G)*qdelay_avg
  end

  # Optional code to calculate the variation on queue delay, which is an
  # Indication of congestion or near congestion.
  qdelay_dev_norm = 0

>   qdelay_dev_norm_th = 1.0

Author: Yes, correct

retransmissions.

qdelay_avg is updated with a slow attack, fast decay EWMA filter as described below. The

> variable qdelay_dev_norm indicates how much the queue delay varies, this is calculated normalized to QDELAY_DEV_NORM. A small margin QDELAY_DEV_NORM/4 is implemented to avoid sensitivity to scheduling jitter.

Reviewer: ```suggestion
variable qdelay_dev_norm indicates how much the queue delay varies,
normalized to QDELAY_DEV_NORM. A small margin QDELAY_DEV_NORM/4
is implemented to avoid sensitivity to scheduling jitter.
```

The variable qdelay_dev_norm_th implements an adaptive threshold that increases the restriction on the ref_wnd growth as well as the ref_wnd_overhead when ref_wnd/MSS is small. This reduces delay and rate varaitions at very low ref_wnd caused by the relatively fast roughly MSS per RTT.

~~~
if (now - last_update_qdelay_avg_time >= min(virtual_rtt,s_rtt)
  # Calculate qdelay_avg
  if (qdelay < qdelay_avg)
    qdelay_avg = qdelay
  else
    qdelay_avg = QDELAY_AVG_G*qdelay + (1.0-QDELAY_AVG_G)*qdelay_avg
  end
  # Calculate qdelay_dev_norm and cap in range [0.0 0.2]

>   tmp = max(0, min(0.2, (qdelay-QDELAY_DEV_NORM/4)/QDELAY_DEV_NORM))

Reviewer: This seems really very random to me to limit the delay samples to 0.2 and reduce them by QDELAY_DEV_NORM/4. If you want to filter out single outliners are are better ways than falsifying all samples!

The SCReAM algorithm can be further improved for a greater rate stability. The outline is to take varaitions in qdelay into consideration, the challenge is to discriminate between delay variations caused by e.g. link layer related scheduling and retransmissions from actual queue delay caused by congestion. The example code below is a simple implementation but more advanced statistical analysis can be considered.

The variable qdelay_dev_norm indicates how much the queue delay varies,
normalized to QDELAY_DEV_NORM. A small margin QDELAY_DEV_NORM/4 is implemented to reduce sensitivity to link layer scheduling jitter and retransmissions. In addition, the metric is capped to avoid that qdelay_dev_norm winds up to very large values in cases of severe congestion.
The variable qdelay_dev_norm_th implements an adaptive threshold that increases the restriction on the ref_wnd growth as well as the ref_wnd_overhead when ref_wnd/MSS is small. This reduces delay and rate variations at very low ref_wnd caused by the relatively fast roughly MSS per RTT.

~~~
  # Additional code to be included in the if-clause above
  # Calculate qdelay_dev_norm_th

>   qdelay_dev_norm_th = max(0.05, 0.1*(1.0-ref_wnd_ratio/0.1))

Author: See modified text

The variable qdelay_dev_norm indicates how much the queue delay varies,
normalized to QDELAY_DEV_NORM. A small margin QDELAY_DEV_NORM/4 is implemented to reduce sensitivity to link layer scheduling jitter and retransmissions. In addition, the metric is capped to avoid that qdelay_dev_norm winds up to very large values in cases of severe congestion. This limit is relative to qdelay_dev_norm_th.  qdelay_dev_norm_th is an adaptive threshold that limits ref_wnd growth and  ref_wnd_overhead and increases if ref_wnd/MSS is small. This reduces delay and rate variations caused by the relatively fast roughly MSS per RTT when the ref_wnd is small. The threshold is limited to a maximum value of 0.1 which is applied when the standard deviation of the delay jitter exceeds the threshold.

~~~
function calculate_qdelay_norm()
  # Calculate qdelay_dev_norm_th. The threshold is 0.1 when ref_wnd > 10*MSS and is gradually
  # decreased to 0.05 when ref_wnd < 10*MSS, the values 0.1 and 0.05 are based on experimentation.
  qdelay_dev_norm_th = max(0.05, 0.1*(1.0-ref_wnd_ratio/0.1))

>   # Calculate qdelay_dev_norm and cap in range [0.0 qdelay_dev_norm_th*1.5]

Author: Hmm, obviously yes!, fixed

* QDELAY_AVG_G (1/4): Exponentially Weighted Moving Average (EWMA) factor for qdelay_avg

The SCReAM algorithm can be further improved for a greater rate stability by taking variations in qdelay into consideration. The goal is to react less to delay variations, caused by e.g. link layer related scheduling and retransmissions, but still be reactive to actual queue delay, caused by congestion. The code below provides a example implementation but more advanced statistical analysis can be considered.

The variable qdelay_dev_norm indicates how much the queue delay varies,

> normalized to QDELAY_DEV_NORM. A small margin QDELAY_DEV_NORM/4 is implemented to reduce sensitivity to link layer scheduling jitter and retransmissions. In addition, the metric is capped to avoid that qdelay_dev_norm winds up to very large values in cases of severe congestion. This limit is relative to qdelay_dev_norm_th.  qdelay_dev_norm_th is an adaptive threshold that limits ref_wnd growth and  ref_wnd_overhead and increases if ref_wnd/MSS is small. This reduces delay and rate variations caused by the relatively fast roughly MSS per RTT when the ref_wnd is small. The threshold is limited to a maximum value of 0.1 which is applied when the standard deviation of the delay jitter exceeds the threshold.

Author: Tried with "This reduces delay and rate variations caused by that the relative increase of ref_wnd is faster when ref_wnd is small"

~~~
function calculate_qdelay_norm()
  # Calculate qdelay_dev_norm_th. The threshold is 0.1 when ref_wnd > 10*MSS and is gradually
  # decreased to 0.05 when ref_wnd < 10*MSS, the values 0.1 and 0.05 are based on experimentation.
  qdelay_dev_norm_th = max(0.05, 0.1*(1.0-ref_wnd_ratio/0.1))

  # Calculate qdelay_dev_norm and cap in range [0.0, qdelay_dev_norm_th*1.5]
  tmp = max(0, min(qdelay_dev_norm_th*1.5, (qdelay-QDELAY_DEV_NORM/4)/QDELAY_DEV_NORM))
  qdelay_dev_norm = (1.0-QDELAY_DEV_AVG_G) * qdelay_dev_norm +
     QDELAY_DEV_AVG_G * tmp
end
~~~

The following variables are used:¶

* qdelay_dev_norm (0): indicates how much the queue delay varies, normalized to QDELAY_DEV_NORM.

* qdelay_dev_norm_th (1.0): an adaptive threshold that limits ref_wnd growth and  ref_wnd_overhead.

> The following constants are used:

Reviewer: ```suggestion

The following constants are used:
```

    # L4S mode
    backoff_t = l4s_alpha / 2

    # Scale down backoff when RTT is high to avoid overreaction to
    # congestion
    backoff_t /= max(1.0, s_rtt/VIRTUAL_RTT)

    if (qdelay < qdelay_target * 0.25)
        # Scale down backoff if close to the last known max reference window
        # This is complemented with a scale down of the reference window increase
        backoff_t *= max(0.25, scl_t)

>         # Optional additional code for increased rate stability

Reviewer: if you use the "if (REDUCE_JITTER == true)" above you also need to use it here

    # L4S mode
    backoff_t = l4s_alpha / 2

    # Scale down backoff when RTT is high to avoid overreaction to
    # congestion
    backoff_t /= max(1.0, s_rtt/VIRTUAL_RTT)

    # Jitter is considered large if the qdelay is larger than qdelay_target/4
    # when L4S is enabled
    if (qdelay < qdelay_target * 0.25)
        # Scale down backoff if close to the last known max reference window
        # This is complemented with a scale down of the reference window increase
        backoff_t *= max(0.25, scl_t)

>         # Optional additional code for increased rate stability

Reviewer: ```suggestion
        # Optional additional code for increased rate stability
        # qdelay_dev_norm is zero if REDUCE_JITTER is false
```

# Reduce increment for small RTTs
tmp_t = min(1.0, s_rtt / VIRTUAL_RTT)
increment_t *= tmp_t

# Apply limit to reference window growth when close to last
# known max value before congestion
increment_t *= max(0.25,scl_t)

# Optional additional code for increased rate stability
# Put a additional restriction on reference window growth if qdelay varies a lot.

> # Better to enforce a slow increase in reference window and get

Author: Yes, fixed

  robustness with less overhead. QUIC {{RFC9000}} overcomes this issue because
  of inherent design.

> * SCReAM has been designed to target 2 marked packets per RTT in steady state when L4S is enabled. There are however a few measures taken in the calculation of the ref_wnd and the target_bitrate, that are necessary to get a stable bitrate and lower queue delay, that make SCReAM settle for a lower number of marked packets per RTT in steady state. The result of this is that SCReAM may get a lower share of the link capacity when competing against e.g. a large file transfer with TCP Prague congestion control.

Reviewer: ```suggestion
* SCReAM has been designed to target two marked packets per RTT in steady state when L4S is enabled.
   However, SCReAM can settle for a lower number of marked packets per RTT in steady state
   due to measures taken in the calculation of the ref_wnd and the target_bitrate
   that are necessary to get a stable bitrate and lower queue delay.
   In those cases SCReAM may get a lower share of the link capacity
   when competing against e.g. a large file transfer with TCP Prague congestion control.
```
